pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,79 +22,309 @@ import (
+			glog.Errorf("Couldn't get object from tombstone %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+		rc, ok = tombstone.Obj.(*api.ReplicationController)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not an rc %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+	}
+	if d := dc.getDeploymentForRC(rc); d != nil {
+		dc.enqueueDeployment(d)
+	}
 }
 
-func (d *DeploymentController) reconcileDeployments() []error {
-	list, err := d.expClient.Deployments(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+// obj could be an *api.Deployment, or a DeletionFinalStateUnknown marker item.
+func (dc *DeploymentController) enqueueDeployment(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
 	if err != nil {
-		return []error{fmt.Errorf("error listing deployments: %v", err)}
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
 	}
-	errs := []error{}
-	for _, deployment := range list.Items {
-		if err := d.reconcileDeployment(&deployment); err != nil {
-			errs = append(errs, fmt.Errorf("error in reconciling deployment %s: %v", deployment.Name, err))
-		}
+
+	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
+	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
+	// ensure that the same deployment is synced for a given RC. When we periodically relist
+	// all deployments there will still be some RC instability. One way to handle this is
+	// by querying the store for all deployments that this deployment overlaps, as well as all
+	// deployments that overlap this deployments, and sorting them.
+	dc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (dc *DeploymentController) worker() {
+	for {
+		func() {
+			key, quit := dc.queue.Get()
+			if quit {
+				return
+			}
+			defer dc.queue.Done(key)
+			err := dc.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing deployment: %v", err)
+			}
+		}()
 	}
-	return errs
 }
 
-func (d *DeploymentController) reconcileDeployment(deployment *extensions.Deployment) error {
-	switch deployment.Spec.Strategy.Type {
+// syncDeployment will sync the deployment with the given key if it has had its expectations fulfilled.
+// This function is not meant to be invoked concurrently with the same key.
+func (dc *DeploymentController) syncDeployment(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing deployment %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	obj, exists, err := dc.dStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("Deployment has been deleted %v", key)
+		return nil
+	}
+	if err != nil {
```

Comment:
This should be before checking exists.
We should be checking exists only if err != nil